airflow/api_fastapi/execution_api/routes/task_instances.py

```diff
@@ -440,3 +462,103 @@ def _is_eligible_to_retry(state: str, try_number: int, max_tries: int) -> bool:
     # max_tries is initialised with the retries defined at task level, we do not need to explicitly ask for
     # retries from the task SDK now, we can handle using max_tries
     return max_tries != 0 and try_number <= max_tries
+
+
+def register_asset_changes(task_instance, task_outlets, outlet_events, asset_type, session):
+    # One task only triggers one asset event for each asset with the same extra.
+    # This tuple[asset uri, extra] to sets alias names mapping is used to find whether
+    # there're assets with same uri but different extra that we need to emit more than one asset events.
+    asset_alias_names: dict[tuple[AssetUniqueKey, frozenset], set[str]] = defaultdict(set)
+    asset_name_refs: set[str] = set()
+    asset_uri_refs: set[str] = set()
+
+    for obj in task_outlets:
+        # Lineage can have other types of objects besides assets
+        if asset_type == "Asset":
+            asset_manager.register_asset_change(
+                task_instance=task_instance,
+                asset=Asset(name=obj.name, uri=obj.uri),
+                extra=outlet_events,
+                session=session,
+            )
+        elif asset_type == "AssetNameRef":
+            asset_name_refs.add(obj.name)
+        elif asset_type == "AssetUriRef":
+            asset_uri_refs.add(obj.uri)
+
+    if asset_type == "AssetAlias":
+        # deserialize to the expected type
+        outlet_events = list(
+            map(
+                lambda event: {**event, "dest_asset_key": AssetUniqueKey(**event["dest_asset_key"])},
+                outlet_events,
+            )
+        )
```

Comment:
Converting it to `AssetUniqueKey` format so that the below access can remain simple and in lines with legacy code. Otherwise Line 502 will complain that we cannot use mutable key for dictionary